Add stub workflow lint template

j3parker opened the pull request.

omsmith commented:
`workflow-templates` isn't actually a thing, right?

j3parker commented:
> `workflow-templates` isn't actually a thing, right?

It is :)

https://help.github.com/en/actions/configuring-and-managing-workflows/sharing-workflow-templates-within-your-organization

omsmith commented on d2l.svg:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<svg xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns="http://www.w3.org/2000/svg" height="32.086mm" width="76.077mm" version="1.1" xmlns:cc="http://creativecommons.org/ns#" xmlns:dc="http://purl.org/dc/elements/1.1/" viewBox="0 0 269.565 113.69169">
+ <g transform="translate(-14.617 -13.648)" fill="#ee7421">
+  <path d="m72.85 57.093c0-13.093-4.0064-21.722-20.027-21.722h-17.1v72.095h17.1c16.021 0 20.027-8.6238 20.027-21.717v-28.656zm-18.484 70.247h-29.272c-5.8543 0-10.477-4.7751-10.477-10.629v-91.043c0-5.8543 4.6224-10.629 10.477-10.629h29.272c34.2 0 39.591 24.802 39.591 41.744v28.966c0 17.095-5.3911 41.591-39.591 41.591"/>
+  <path d="m163.59 127.34h-50.373c-6.1648 0-10.171-4.1591-10.171-10.166 0-16.947 7.5495-27.266 17.41-35.895 14.173-12.325 32.041-19.564 32.041-34.047 0-7.0863-4.0064-14.325-14.478-14.325-6.1648 0-10.324 1.5424-13.709 3.5432-4.4696 2.4639-6.628 5.0856-10.634 5.0856-5.2334 0-10.009-4.3168-10.009-10.013 0-2.7744 0.76875-5.5488 3.6959-7.86 8.3183-6.6231 18.948-10.013 31.425-10.013 25.418 0 34.968 18. ...
+  <path d="m245.55 127.34h-49.451c-5.5439 0-10.013-4.4696-10.013-10.166v-92.896c0-5.8494 4.6224-10.629 10.477-10.629 6.0071 0 10.629 4.78 10.629 10.629v82.882h38.359c5.5439 0 10.166 4.4647 10.166 10.013 0 5.6966-4.6224 10.166-10.166 10.166"/>
+  <path d="m261.86 19.541h-4.7209v-1.5868h11.329v1.5868h-4.7258v12.576h-1.8824v-12.576z"/>
+  <path d="m270.08 17.955h2.5822l4.4647 11.901 4.4794-11.901h2.5773v14.163h-1.7839v-11.783h-0.0394l-4.4203 11.783h-1.6114l-4.4203-11.783h-0.0394v11.783h-1.7888v-14.163z"/>
+ </g>
+</svg>
```
```suggestion
<svg xmlns="http://www.w3.org/2000/svg" height="121.27" width="287.535" viewBox="0 0 269.565 113.692"><g fill="#ee7421"><path d="M58.233 43.445c0-13.093-4.006-21.722-20.027-21.722h-17.1v72.095h17.1c16.021 0 20.027-8.624 20.027-21.717V43.445zm-18.484 70.247H10.477C4.623 113.692 0 108.917 0 103.063V12.02C0 6.166 4.622 1.391 10.477 1.391h29.272c34.2 0 39.591 24.802 39.591 41.744v28.966c0 17.095-5.391 41.591-39.591 41.591M148.973 113.692H98.6c-6.165 0-10.171-4.16-10.171-10.166 0-16.947 7.55-27.266 17.41-35.895 14.173-12.325 32.041-19.564 32.041-34.047 0-7.086-4.006-14.325-14.478-14.325-6.165 0-10.324 1.542-13.709 3.543-4.47 2.464-6.628 5.086-10.634 5.086-5.233 0-10.009-4.317-10.009-10.013 0-2.775.769-5.549 3.696-7.86C101.064 3.392 111.694.002 124.17.002c25.418 0 34.968 18.332 34.968 33.584 0 32.352-38.975 37.284-47.598 59.928h37.432c5.702 0 10.166 4.465 10.166 10.013 0 5.696-4.465 10.166-10.166 10.166M230.933 113.692h-49.451c-5.544 0-10.013-4.47-10.013-10.166V10.63c0-5.85 4.622-10.629 10.477-10.629 6.007 0 10.629 4.78 10.629 10.629v82.882h38.359c5.544 0 10.166 4.465 10.166 10.013 0 5.697-4.622 10.166-10.166 10.166M247.243 5.893h-4.72V4.306h11.328v1.587h-4.726v12.576h-1.882V5.893zM255.463 4.307h2.582l4.465 11.901 4.48-11.901h2.577V18.47h-1.784V6.687h-.04l-4.42 11.783h-1.611l-4.42-11.783h-.04V18.47h-1.789V4.307z"/></g></svg>
```

Ran through svgomg. Dunno if this is our branding logo if we shouldn't though.

j3parker replied:
Heh I grabbed this one from Wikipedia :)

omsmith commented:
It's unclear to me if `filePatterns` is desirable here. i.e. if it restricts you from using it if you don't have a matching file, or if it actually suggests it if you have a matching file.

j3parker commented:
I think it restricts it being suggested/usable via the UI. So I understood it as we could have a `.github/workflows/*.yml` pattern but then we'd never suggest this workflow to repos without workflows. I think it makes sense to always suggest it, because the suggestion shows up on the page for creating workflows (if you're already there, we probably want to nag you about this).

j3parker commented on actions/workflow-lint/action.yml:
```diff
@@ -0,0 +1,5 @@
+name: D2L GitHub Action Workflow Linter
```
**Why a separate action?** we could probably do it just with a `run:` step, but then because this template will get copy-pasted into repos we are asking for drift.

**Why define the action in this repo?** I don't feel strongly about this. This action is meant to be used with this template file, so I felt it maybe didn't make sense to put in it's own repo. It's the kind of action that I don't think is going to be (directly) useful outside of our org (the way I intend to write it).

j3parker commented on workflow-templates/workflow-lint.yml:
```diff
@@ -0,0 +1,17 @@
+      - '.github/workflows/*.yml'
+  schedule:
+    # Noon (EST) once a month
+    - cron: '0 4 1 * *'
```
Why also have a schedule?
Plenty of repos won't habitually update their workflows. If we do this we can pull failed workflow runs from the API at some point.

Monthly might be a long time thought. Thoughts about daily?

E.g. say you add a new lint here, you'd need to wait until this fired again to see for sure which repos failed the lint.

daryl-mcmillan replied:
Monthly is a long time, but maybe daily is a lot?  Weekly?

j3parker replied:
Might be hard with cron 😬

daryl-mcmillan replied:
The last `*` is day of the week: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/crontab.html#tag_20_25_07

daryl-mcmillan replied:
Ah, you got it

j3parker replied:
https://slack.d2l.dev/emoji/doh

j3parker commented:
Moved the action to another branch. @omsmith thought this would be better because it will reduce the amount of stuff that needs to be cloned to use an action. I'm a bit nervous about "hiding" the code, but I added a link in the README to compensate.
